proof/refine/ARM/Finalise_R.thy

```diff
@@ -4016,6 +4013,81 @@ lemma unbind_maybe_notification_corres:
   apply (clarsimp simp: valid_obj'_def valid_ntfn'_def split: option.splits)
   done
 
+lemma update_sched_context_corres:
+  "\<lbrakk>\<forall>sc n sc'. sc_relation sc n sc' \<longrightarrow> sc_relation (f sc) n (f' sc');
+    \<forall>sc. sc_replies sc = sc_replies (f sc); \<forall>sc'. objBits sc' = objBits (f' sc');
+    \<forall>sc'. scReply (f' sc') = scReply sc'\<rbrakk> \<Longrightarrow>
+   corres dc
+     (sc_at ptr and pspace_aligned and pspace_distinct)
+     (ko_at' sc' ptr)
+     (update_sched_context ptr f)
+     (setSchedContext ptr (f' sc'))"
+  apply (rule_tac Q="sc_at' ptr" in corres_cross_add_guard)
+   apply (fastforce dest!: state_relationD sc_at_cross simp: obj_at'_def)
+  apply (rule_tac Q="sc_obj_at (objBits sc' - minSchedContextBits) ptr" in corres_cross_add_abs_guard)
+   apply (fastforce dest!: state_relationD ko_at'_cross)
+  apply (rule corres_guard_imp)
+    apply (rule_tac P="sc_obj_at (objBits sc' - minSchedContextBits) ptr"
+               and n1="objBits sc' - minSchedContextBits"
+                         in monadic_rewrite_corres[OF _ update_sched_context_rewrite])
+    apply (rule corres_symb_exec_l)
+       apply (rule corres_guard_imp)
+         apply (rule_tac P="\<lambda>s. kheap s ptr = Some (kernel_object.SchedContext sc (objBits sc' - minSchedContextBits))"
+                     and P'="ko_at' sc' ptr"  in corres_inst)
+         apply (rule_tac n="objBits sc' - minSchedContextBits" in setSchedContext_no_stack_update_corres)
+            apply simp+
+      apply (wpsimp wp: get_sched_context_exs_valid simp: is_sc_obj_def obj_at_def)
+       apply (rename_tac ko; case_tac ko; simp)
+      apply simp
+     apply (wpsimp simp: obj_at_def is_sc_obj_def)+
+    apply (wpsimp wp: get_sched_context_no_fail)
+   apply (clarsimp simp: obj_at_def is_sc_obj_def)
+  apply simp
+  done
+
+lemma sched_context_maybe_unbind_ntfn_corres:
+  "corres dc
+     (invs and ntfn_at ntfn_ptr)
+     (invs' and ntfn_at' ntfn_ptr)
```

Comment:
You could use some sort of cross lemma to give you` ntfn_at'` in the concrete guard

Reply:
Is it good practice to have differing guards and using cross lemmas to get what we need for the sake of minimality?

Reply:
Big time